pkg/api/mcapi.go

```diff
@@ -48,89 +50,132 @@ func NewClient() *Client {
+		api_client, err = common.Login(myclient.Ctx, config)
+
+		if err == nil && api_client != nil {
+			myclient.apiClient = api_client
+			configuration := api_client.GetConfig()
+			myclient.SessionKey = configuration.DefaultHeader["sessionKey"]
+		}
+		// if we have a new session key, login was successful and we can leave
+		// if not, try the next controller IP
+		if myclient.SessionKey != "" {
+			break
+		}
 	}
-	_, _, err := client.apiClient.DefaultApi.LogoutGet(client.Ctx).Execute()
 	return err
 }
 
-// SessionValid : Determine if a session is valid, if not a login is required
-// Makes the 'show controller-date' API call to validate the session is still valid
-//
-// Deprecated: This function will be made redundant by retry logic in a future update
-func (client *Client) SessionValid(addr, username string) bool {
+// Logout: Called to log out of the storage controller API
+func (client *Client) Logout() error {
 	if client.Ctx == nil {
-		return false
+		client.Ctx = context.Background()
 	}
-	logger := klog.FromContext(client.Ctx)
-
-	// addr may include protocol and ip address or hostname, or only ip address
-	ipaddress := common.GetAddress(addr)
-
-	if client.Addr == ipaddress && client.Username == username {
-		if client.SessionKey == "" {
-			logger.V(2).Info("session invalid", "sessionkey", client.SessionKey)
-			return false
-		}
+	_, _, err := client.apiClient.DefaultApi.LogoutGet(client.Ctx).Execute()
+	if err == nil {
+		client.SessionKey = ""
 	}
-
-	//run an API call to test that the session is still valid
-	_, _, err := client.apiClient.DefaultApi.ShowControllerDateGet(client.Ctx).Execute()
-	return err == nil
+	return err
 }
 
 // InitSystemInfo: Retrieve and store system information for this client
 func (client *Client) InitSystemInfo() error {
 
-	err := AddSystem(client.Addr, client)
+	err := AddSystem(client.CurrentAddr, client)
 	if err != nil {
-		return fmt.Errorf("unable to add system info for ip (%s) ", client.Addr)
+		return fmt.Errorf("unable to add system info for ip (%s) ", client.CurrentAddr)
 	}
 
-	client.Info, err = GetSystem(client.Addr)
+	client.Info, err = GetSystem(client.CurrentAddr)
 	if err == nil {
 		_ = Log(client.Info)
 	}
 
 	return err
 }
 
+type ResponseType interface {
+	GetStatus() []client.StatusResourceInner
+}
+
+// ExecuteWithFailover: Retry wrapper for the Execute functions of the openapi generated client library
+// If the initial request fails, attempts to login on the secondary controller
+func ExecuteWithFailover[R ResponseType](executeFunc func() (R, *http.Response, error), client *Client) (R, *common.ResponseStatus, *http.Response, error) {
+	logger := klog.FromContext(client.Ctx)
+	logger.V(4).Info("Execute with retry...")
+	response, httpRes, err := executeFunc()
+	if err == nil {
+		status := response.GetStatus()
+		commonStatus := CreateCommonStatus(logger, &status)
+		return response, commonStatus, httpRes, err
+	}
+	if len(client.Addrs) > 1 {
```

Comment:
I think you should check the MC return code against a list here so that you don't retry certain kinds of commands that aren't worth retrying, e.g. a command that fails with "insufficient space in pool", etc.  I'd also add a way for the high-level client to modify the list on each call.  This was something that we needed to do for the SRA where commands can be time-critical and we couldn't afford to waste time on retries that were not going to succeed.  But if you don't foresee that problem in the CSI driver, that could be a future addition.